core/src/main/java/hudson/PluginFirstClassLoader2.java

```diff
@@ -0,0 +1,136 @@
+package hudson;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.util.CompoundEnumeration;
+import java.io.IOException;
+import java.net.URL;
+import java.util.Enumeration;
+import java.util.Objects;
+import jenkins.util.URLClassLoader2;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Class loader that consults the plugin's {@code WEB-INF/lib/*.jar} and {@code WEB-INF/classes}
+ * directories and the Jenkins core class loader (in that order).
+ *
+ * <p>To use this class loader, set {@code pluginFirstClassLoader} to {@code true} in the {@code
+ * maven-hpi-plugin} configuration.
+ *
+ * @author Basil Crow
+ * @since TODO
+ */
+@Restricted(NoExternalUse.class)
+public class PluginFirstClassLoader2 extends URLClassLoader2 {
+    static {
+        registerAsParallelCapable();
+    }
+
+    public PluginFirstClassLoader2(@NonNull URL[] urls, @NonNull ClassLoader parent) {
+        super(Objects.requireNonNull(urls), Objects.requireNonNull(parent));
+    }
+
+    /**
+     * Load the class with the specified binary name. This method searches for classes in the
+     * following order:
+     *
+     * <ol>
+     *   <li>
+     *       <p>Invoke {@link #findLoadedClass(String)} to check if the class has already been
+     *       loaded.
+     *   <li>
+     *       <p>Invoke {@link #findClass(String)} to find the class.
+     *   <li>
+     *       <p>Invoke {@link #loadClass(String)} on the parent class loader.
```

Comment:
Matches the implementation here but not the PR description

> The implementation delegates first to the system class loader, then to the plugin class loader, then to the parent (i.e. core) class loader.

unless I am missing something?

Reply:
> unless I am missing something

You are missing the review by @jtnord during which I was asked to remove the delegation to the system class loader.

Reply:
https://github.com/jenkinsci/jenkins/pull/5970#discussion_r765713610

Reply:
So the PR description is out of date and should be edited?

Reply:
> So the PR description is out of date

Correct